presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java

```diff
@@ -535,11 +591,39 @@ protected void invalidateTable(String databaseName, String tableName)
             String tableName,
             Map<Column, Domain> partitionPredicates)
     {
+        if (partitionVersioningEnabled) {
+            List<PartitionNameWithVersion> partitionNamesWithVersion = getPartitionNamesWithVersionByFilter(databaseName, tableName, partitionPredicates);
+            List<String> result = partitionNamesWithVersion.stream().map(PartitionNameWithVersion::getPartitionName).collect(toImmutableList());
+            partitionNamesWithVersion.forEach(partitionNameWithVersion -> invalidateStalePartition(partitionNameWithVersion, databaseName, tableName));
+            return result;
+        }
         return get(
                 partitionFilterCache,
                 partitionFilter(databaseName, tableName, partitionPredicates));
     }
 
+    @Override
+    public List<PartitionNameWithVersion> getPartitionNamesWithVersionByFilter(
+            String databaseName,
+            String tableName,
+            Map<Column, Domain> partitionPredicates)
+    {
+        return delegate.getPartitionNamesWithVersionByFilter(databaseName, tableName, partitionPredicates);
+    }
+
+    private void invalidateStalePartition(PartitionNameWithVersion partitionNameWithVersion, String databaseName, String tableName)
```

Comment:
We probably could add some unit tests for this by calling `getPartitionNamesByFilter` in `TestCachingHiveMetastore` since this is the core logic for this feature.

Reply:
Added `testCachingWithPartitionVersioning()` to `TestCachingHiveMetastore`